### ISSUE TITLE
1600 lines of create tables - take up increasing amount of CPU resource.

### ISSUE THREAD

**Issue by biglobe** [cockroachdb/cockroach#31421]
Hi, I am getting started with Cockroach and coming from SQL Server background.

I am currently migrating the tables from SQL into CR using a program written in C# with Npgsql. The code loads all table in SQL and for each table, the columns are read and the create table statement is generated. This statement is called through Npgsql. One call per create statement.

I have 3 nodes, each running in separate VM, 4 CPUs, 8GB ram.

Initially, CR runs ok.
Resource on each CPU is < 10%.
Memory is considered stable. Raising from 600MB to 900MB until completion.
However after about 500 tables are created, CPU resource goes up to 100%, Admin UI stops responding but tables are still created. Eventually create table slows to about 1.5s per create statement.

I'm not sure why CPU resource is increasingly being eaten up, but it shouldnt be happening. I suspect transactional problems, but i'm not explicitly declaring transactions.

Appreciate if there is guidance to figure this out, otherwise please use the information to improve CR.

![image](https://user-images.githubusercontent.com/4609085/46992788-648deb00-d13e-11e8-9d1d-abf287e7b8a0.png)

**Comment by vivekmenezes** [comment]
@biglobe thanks for reporting this issue. We unfortunately do not support schemas with thousands of tables well today. The relevant issue is #28376 . We have plans to fix this in release 2.2 coming out in the spring.

**Comment by biglobe** [comment]
@vivekmenezes thanks for the response. #28376 is an issue on large numbers of databases, whereas mine is about large number of tables. Are they really related?

For the record, the problem also occurs for large number of inserts. x00,000 range of records. Slowly down occurs at about 3000 records in a single statement and into a single table.

Cheers!

**Comment by vivekmenezes** [comment]
@biglobe yeah this is the same issue as that one as far as a large number of tables is concerned. But I think your issues is a bit different and is likely an admin UI issue. I'll investigate and get back to you.

As far as multi-row INSERT performance is concerned you might find https://www.cockroachlabs.com/blog/multi-row-dml/ a valuable read.

Thank you! I'm keeping this issue open.

**Comment by vilterp** [comment]
@vivekmenezes when we discussed this earlier I thought it was an issue with the databases page specifically, which is what I'm looking into now, since its data fetching is inefficient and could take up a lot of CPU (#22941). However, I don't see in this report that @biglobe even went to the databases page, so I don't think that endpoint is to blame for the high CPU usage.

It's probably some other issue, like the ones you linked to, inherent to a large number of databases/tables, maybe at the gossip level. Taking a CPU profile would let us know for sure. @biglobe, you can take a CPU profile by going to `http(s)://<your-cluster-address>/debug/pprof/ui/profile/`.

**Comment by biglobe** [comment]
@vilterp i will capture and post it later.

Anyway I am encountering another performance issue here. I am doing mass delete from existing tables, but only 10 lines of delete statements at a time. Each table has at most 2000 records and the tables have the same schema. No other commands are running at the same time.

At first I used `DELETE` statement which took a really long time and later I tried it with 'TRUNCATE' which took about 60% of the time, which is still too long. I later found that out of the 10 statements, if I run the statements individually, there will be 1 or 2 of the tables that are the culprits that block the run. The funny thing is that it is always the same tables that block. The individual table will finally complete their run after 4 mins or so.

2.1-RC further reduces the time by another 30% or so, but I am encountering much more "Attempt to read past end of stream errors" from Npgsql. However, there is no further exception messages that could help decipher the underlying problem.

Not sure if this can still be attributed to the same problem, otherwise let me know. I will open another issue.

Wish I could help more and thanks for listening out!
